File: backend/app/api/routers.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import require_admin, require_permission
from app.models.usuario import Usuario
from app.schemas.common import PaginatedResponse
from app.schemas.router import (
    RouterCreate,
    RouterResponse,
    RouterTestConnectionResponse,
    RouterUpdate,
)
from app.services import routers as routers_service
from app.services.mikrotik import MikroTikService

router = APIRouter(prefix="/routers", tags=["Routers"])


@router.get("/", response_model=PaginatedResponse[RouterResponse])
async def list_routers(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str | None = None,
    is_active: bool | None = None,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers")),
):
    """List routers (requires read permission on routers module)"""
    return await routers_service.list_routers(db, page, page_size, search, is_active)


@router.get("/{router_id}", response_model=RouterResponse)
async def get_router(
    router_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers")),
):
    """Get router by ID (requires read permission on routers module)"""
    return await routers_service.get_router(db, router_id)


@router.post("/", response_model=RouterResponse, status_code=201)
async def create_router(
    data: RouterCreate,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers", require_write=True)),
):
    """Create new router (requires write permission on routers module)"""
    return await routers_service.create_router(db, data)


@router.put("/{router_id}", response_model=RouterResponse)
async def update_router(
    router_id: uuid.UUID,
    data: RouterUpdate,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers", require_write=True)),
):
    """Update router (requires write permission on routers module)"""
    return await routers_service.update_router(db, router_id, data)


@router.delete("/{router_id}", status_code=204)
async def delete_router(
    router_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers", require_write=True)),
):
    """Delete router (requires write permission on routers module)"""
    await routers_service.delete_router(db, router_id)


@router.post("/{router_id}/test-connection", response_model=RouterTestConnectionResponse)
async def test_router_connection(
    router_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers")),
):
    """
    Test connection to MikroTik router

    Attempts to connect to the router and retrieve system identity
    to verify credentials and connectivity.
    """
    router_obj = await routers_service.get_router(db, router_id)
    password = routers_service.decrypt_router_password(router_obj)

    mikrotik_service = MikroTikService(
        host=router_obj.ip,
        username=router_obj.usuario,
        password=password,
        port=router_obj.puerto,
        ssl=router_obj.ssl,
    )

    return await mikrotik_service.test_connection()


@router.post("/{router_id}/deactivate", response_model=RouterResponse)
async def deactivate_router(
    router_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers", require_write=True)),
):
    """Deactivate router (requires write permission on routers module)"""
    return await routers_service.deactivate_router(db, router_id)


@router.post("/{router_id}/activate", response_model=RouterResponse)
async def activate_router(
    router_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers", require_write=True)),
):
    """Activate router (requires write permission on routers module)"""
    return await routers_service.activate_router(db, router_id)


@router.get("/{router_id}/next-available-ip")
async def get_next_available_ip(
    router_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers")),
):
    """Get next available IP address from router's CIDR ranges"""
    ip_address = await routers_service.get_next_available_ip(db, router_id)
    return {"ip_address": ip_address}


@router.get("/{router_id}/check-ip/{ip_address}")
async def check_ip_available(
    router_id: uuid.UUID,
    ip_address: str,
    exclude_contrato_id: uuid.UUID | None = Query(None),
    db: AsyncSession = Depends(get_db),
    _current_user: Usuario = Depends(require_permission("routers")),
):
    """Check if IP address is available for assignment (Admin and Operador only)"""
    return await routers_service.check_ip_available(db, router_id, ip_address, exclude_contrato_id)

File: frontend/src/api/routers.ts
import type { PaginatedResponse } from "@/types/common";
import type {
  Router,
  RouterCreate,
  RouterTestConnectionResponse,
  RouterUpdate,
} from "@/types/router";
import api from "./axios";

export async function getRouters(params: {
  page?: number;
  page_size?: number;
  search?: string;
  is_active?: boolean;
}): Promise<PaginatedResponse<Router>> {
  const { data } = await api.get<PaginatedResponse<Router>>("/routers/", { params });
  return data;
}

export async function getRouter(id: string): Promise<Router> {
  const { data } = await api.get<Router>(`/routers/${id}`);
  return data;
}

export async function createRouter(router: RouterCreate): Promise<Router> {
  const { data } = await api.post<Router>("/routers/", router);
  return data;
}

export async function updateRouter(id: string, router: RouterUpdate): Promise<Router> {
  const { data } = await api.put<Router>(`/routers/${id}`, router);
  return data;
}

export async function deleteRouter(id: string): Promise<void> {
  await api.delete(`/routers/${id}`);
}

export async function testConnection(id: string): Promise<RouterTestConnectionResponse> {
  const { data } = await api.post<RouterTestConnectionResponse>(
    `/routers/${id}/test-connection`
  );
  return data;
}

export async function deactivateRouter(id: string): Promise<Router> {
  const { data } = await api.post<Router>(`/routers/${id}/deactivate`);
  return data;
}

export async function activateRouter(id: string): Promise<Router> {
  const { data } = await api.post<Router>(`/routers/${id}/activate`);
  return data;
}

export async function getNextAvailableIp(routerId: string): Promise<{ ip_address: string }> {
  const { data } = await api.get<{ ip_address: string }>(
    `/routers/${routerId}/next-available-ip`
  );
  return data;
}

export async function checkIpAvailable(
  routerId: string,
  ipAddress: string,
  excludeContratoId?: string
): Promise<{
  available: boolean;
  message: string;
  contrato_numero: string | null;
}> {
  const params = excludeContratoId ? { exclude_contrato_id: excludeContratoId } : {};
  const { data } = await api.get(
    `/routers/${routerId}/check-ip/${ipAddress}`,
    { params }
  );
  return data;
}
